### PR TITLE
Bugfix/saml load balancer context

### PR DESCRIPTION
**Pull request by wtritch**
After changing Spring to handle X-Forwarded headers set by the AWS ELB our QA environment appeared to support SAML authentication over SSL terminated at the ELB.
Unfortunately, this doesn't appear to work 100% of the time, leading to occasional random errors where the reporting app does not appear to recognize/parse the X-Forwarding headers:
```
Fairways-MacBook-Pro-6:webapp willtritchler$ kubectl logs reporting-deployment-3046325867-5wb55 | grep recipient
2017-08-02 00:47:48.465 ERROR 1 --- [nio-8080-exec-6] o.o.c.b.decoding.BaseSAMLMessageDecoder  : SAML message intended destination endpoint 'https://awsqa-reporting.sbacdw.org/saml/SingleLogout' did not match the recipient endpoint 'http://awsqa-reporting.sbacdw.org/saml/SingleLogout'
org.opensaml.xml.security.SecurityException: SAML message intended destination endpoint did not match recipient endpoint
2017-08-02 15:49:44.672 ERROR 1 --- [io-8080-exec-11] o.o.c.b.decoding.BaseSAMLMessageDecoder  : SAML message intended destination endpoint 'https://awsqa-reporting.sbacdw.org/saml/SSO' did not match the recipient endpoint 'http://awsqa-reporting.sbacdw.org/saml/SSO'
2017-08-02 17:19:35.393 ERROR 1 --- [nio-8080-exec-9] o.o.c.b.decoding.BaseSAMLMessageDecoder  : SAML message intended destination endpoint 'https://awsqa-reporting.sbacdw.org/saml/SingleLogout' did not match the recipient endpoint 'http://awsqa-reporting.sbacdw.org/saml/SingleLogout'
org.opensaml.xml.security.SecurityException: SAML message intended destination endpoint did not match recipient endpoint
2017-08-02 17:23:26.431 ERROR 1 --- [nio-8080-exec-2] o.o.c.b.decoding.BaseSAMLMessageDecoder  : SAML message intended destination endpoint 'https://awsqa-reporting.sbacdw.org/saml/SSO' did not match the recipient endpoint 'http://awsqa-reporting.sbacdw.org/saml/SSO'
2017-08-02 17:24:22.190 ERROR 1 --- [io-8080-exec-12] o.o.c.b.decoding.BaseSAMLMessageDecoder  : SAML message intended destination endpoint 'https://awsqa-reporting.sbacdw.org/saml/SingleLogout' did not match the recipient endpoint 'http://awsqa-reporting.sbacdw.org/saml/SingleLogout'
org.opensaml.xml.security.SecurityException: SAML message intended destination endpoint did not match recipient endpoint
2017-08-02 17:34:47.353 ERROR 1 --- [nio-8080-exec-1] o.o.c.b.decoding.BaseSAMLMessageDecoder  : SAML message intended destination endpoint 'https://awsqa-reporting.sbacdw.org/saml/SingleLogout' did not match the recipient endpoint 'http://awsqa-reporting.sbacdw.org/saml/SingleLogout'
org.opensaml.xml.security.SecurityException: SAML message intended destination endpoint did not match recipient endpoint
2017-08-02 17:40:42.760 ERROR 1 --- [nio-8080-exec-6] o.o.c.b.decoding.BaseSAMLMessageDecoder  : SAML message intended destination endpoint 'https://awsqa-reporting.sbacdw.org/saml/SSO' did not match the recipient endpoint 'http://awsqa-reporting.sbacdw.org/saml/SSO'
```

This PR changes SAML to wrap requests in an LPRequestWrapper class, which provides configured scheme, host, port, etc values.

### CODE REVIEW

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlSettings.java [hunk 1]
```diff
@@ -13,7 +13,7 @@
     private String privateKeyEntryPassword;
     private String idpMetadataUrl;
     private String spEntityId;
-    private String spEntityBaseUrl;
```
Comment (wtritch) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlSettings.java]: We shouldn't need this setting any more, the metadata generator will use the servlet request to craft this value which will respond with the configured scheme, host, etc.

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlSettings.java [hunk 2]
```diff
@@ -82,13 +82,93 @@ public void setSpEntityId(final String spEntityId) {
...
+    /**
+     * SAML configuration when server is hosted behind a load balancer or reverse proxy.
+     */
+    public static class LoadBalancedSettings {
+        private boolean enabled = false;
+        private String scheme = "https";
+        private String hostname;
+        private String contextPath = "/";
+        private int port;
+        private boolean includeServerPort = false;
```
Comment (wtritch) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlSettings.java]: By default load-balancing is disabled.

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlWebSecurityConfiguration.java [hunk 3]
```diff
@@ -150,8 +152,20 @@ public SAMLAuthenticationProvider samlAuthenticationProvider() {
...
+        }
+
+        final SAMLContextProviderLB contextProvider = new SAMLContextProviderLB();
+        contextProvider.setScheme(lbSettings.getScheme());
+        contextProvider.setServerName(lbSettings.getHostname());
+        contextProvider.setContextPath(lbSettings.getContextPath());
+        contextProvider.setServerPort(lbSettings.getPort());
+        contextProvider.setIncludeServerPortInRequestURL(lbSettings.isIncludeServerPort());
+
+        return contextProvider;
```
Comment (wtritch) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlWebSecurityConfiguration.java]: If load balancing is disabled, use the same SAMLContextProviderImpl implementation we were using before.  If enabled, use a SAMLContextProviderLB with overriding scheme, host, etc values.

File: webapp/src/main/resources/application.yml [hunk 4]
```diff
@@ -12,6 +12,15 @@ saml:
...
   idp-metadata-url: ${rdw.reporting.idp}
   sp-entity-id: ${rdw.reporting.entityId}
+  # SAML settings when hosting behind a load balancer or reverse proxy
+  load-balance:
+    enabled: false
+    scheme: https
+    hostname:
+    context-path: /
+    port: 8080
+    include-server-port: false
```
Comment (wtritch) [inline webapp/src/main/resources/application.yml]: We ran into an issue when deploying TDS where nobody knew all the properties that were available to be configured.  I added some default values in here, but left load-balancing disabled by default.

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlSettings.java [hunk 5]
```diff
@@ -13,7 +13,7 @@
     private String privateKeyEntryPassword;
     private String idpMetadataUrl;
     private String spEntityId;
-    private String spEntityBaseUrl;
+    private LoadBalancedSettings loadBalance = new LoadBalancedSettings();
```
Comment (malaffoon) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlSettings.java]: nit: these are really ProxySettings, just so happens that's a load balancer for our aws environments

File: webapp/src/main/resources/application.yml (hunk 4 above)
Comment (malaffoon) [inline webapp/src/main/resources/application.yml]: Hmm, fine for now. But i lean towards the Spring philosophy: keep the properties files as clean and stripped down as possible while providing a separate Properties Appendix.

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlSettings.java (hunk 1 above)
Comment (malaffoon) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlSettings.java]: Don't forget to test that assertion when we deploy to awsqa.

Comment (wtritch) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/security/saml/SamlSettings.java]: We'll test that real quick =)

File: webapp/src/main/resources/application.yml (hunk 4 above)
Comment (wtritch) [inline webapp/src/main/resources/application.yml]: I like the idea of a Properties Appendix.  When're we going to make one of those? =)

**Pull request merged**